### DISCUSSION TITLE
how to accept a new peer connection based on a specific requirement in Gossipsub

### DISCUSSION THREAD

**Discussion by revoltez** [libp2p/rust-libp2p#3463] [Q&A]
lets say i have a list of public keys (not the same type used in libp2p) and i only want to accept a connection request from some peer only if he posses that public (lets say he sign some message and send a request for connection and i verify that indeed he posses that key i add him, if not the connection is cancelled)
i tried to use identify but i didn't knew how.

**Comment by thomaseizinger** [comment]
You will likely have to create your own protocol for this that carries the custom message back and forth. Look at `libp2p-request-response`.

Now, because a message exchange is required, you cannot guarantee that the connection isn't already used while your "verification" is still going on.

What you could do is create two instances of `Swarm` in your application:

- One with just your verification behaviour. Any peer can connect to this one and prove their identity.

- The second `Swarm` listens on a different port and contains your actual networking behaviours in addition to an "allow-list" behaviour (I am assuming https://github.com/libp2p/rust-libp2p/pull/3254 is merged for this proposal). Said allow-list is populated from peers that verified themselves via the first `Swarm`. Connections from any peers not verified are immediately closed.

**Reply by mxinden** [reply]
Third option that comes to my mind is using the `Toggle` and dynamically activate the remaining `NetworkBehaviour` implementations ones your custom behaviour gives the green light.

**Reply by thomaseizinger** [reply]
> Third option that comes to my mind is using the `ToggleBehaviour` and dynamically activate the remaining `NetworkBehaviour` implementations ones your custom behaviour gives the green light.

That is a really neat idea too!

**Reply by thomaseizinger** [reply]
Currently, `Toggle` only allows to set the state on creation so that would need a small patch to allow toggling at runtime.

**Reply by revoltez** [reply]
@thomaseizinger  why not make something to allow only a specific list of multiaddr that could be accepted in gossipsub (would be much easier and cleaner than creating a different network behviour for new users)? are there any good examples on how to use Toggle?

**Reply by mxinden** [reply]
> so that would need a small patch to allow toggling at runtime.

Wouldn't the existing `Toggle::as_mut` suffice?

**Reply by thomaseizinger** [reply]
> would be much easier and cleaner than creating a different network behviour

I disagree. In general, composition of multiple smaller components tends to be "cleaner" if "clean" is characterized by less code duplication and ease of understanding of the existing components.

If we start to add unrelated connection management logic to each `Networkbehaviour`, we will soon have a plethora of configuration options everywhere that look and behave the same, yet are duplicated and not uniform instead of having it encapsulated in a single component. It might serve your particular usecase well but as library maintainers, we need to balance the design to serve many other usecases too.

> why not make something to allow only a specific list of multiaddr that could be accepted in gossipsub

How would this solve your problem? The multiaddr is not cryptographically-secure. I thought you wanted to manage connections based on public keys?

> are there any good examples on how to use Toggle?

Thinking about it in more detail, I don't think `Toggle` is a good fit here because it would require to structure the application into two runtime phases: First only login and later allow access. Once the `NetworkBehaviour` is activated (via `Toggle`), any peer can connect to it, not just the verified ones. You need an "allow-list" to dynamically manage the connections.

I still stand by my original proposal. Think about it like a website where you have some publicly accessible routes and some that you need to be logged in for.

1. One instance of `Swarm` contains the `NetworkBehaviour` with all the services that unverified users can perform. Unless you have any other services to offer, the only thing a peer can do here is to "login", i.e. using your custom public keys and verification logic. This will require a custom protocol, look at `libp2p-request-response` for that.
2. A second instance of `Swarm` contains all the "gated" services, i.e. the ones you need to be logged in for. With https://github.com/libp2p/rust-libp2p/pull/3254 in place (it will merge soon), it is trivial to create an "allow-list" `NetworkBehaviour` that only allows connections from a list of `PeerId`s.
3. Every time a peer successfully "logs in" on the first `Swarm`, you add their `PeerId` to the allow list of the second `Swarm`.

> > so that would need a small patch to allow toggling at runtime.
> 
> Wouldn't the existing `Toggle::as_mut` suffice?

I only went after the documentation when I wrote this which seems to be outdated: https://github.com/libp2p/rust-libp2p/blob/577e8f95478785f7b2de20605e7ca0d19821709d/swarm/src/behaviour/toggle.rs#L39
